noir/noir-repo/compiler/noirc_frontend/src/hir/def_map/mod.rs

```diff
@@ -217,20 +217,66 @@ impl CrateDefMap {
+
+                    interner.get_all_globals().iter().for_each(|global_info| {
+                        interner.global_attributes(&global_info.id).iter().for_each(|attr| {
+                            if let SecondaryAttribute::Abi(tag) = attr {
+                                if let Some(tagged) = outputs.globals.get_mut(tag) {
+                                    tagged.push(global_info.id);
+                                } else {
+                                    outputs.globals.insert(tag.to_string(), vec![global_info.id]);
+                                }
+                            }
+                        });
+                    });
+
+                    module.type_definitions().for_each(|id| {
+                        if let ModuleDefId::TypeId(struct_id) = id {
+                            interner.struct_attributes(&struct_id).iter().for_each(|attr| {
+                                if let SecondaryAttribute::Abi(tag) = attr {
+                                    if let Some(tagged) = outputs.structs.get_mut(tag) {
+                                        tagged.push(struct_id);
+                                    } else {
+                                        outputs.structs.insert(tag.to_string(), vec![struct_id]);
+                                    }
+                                }
+                            });
+                        }
+                    });
+
+                    let mut outputs =
+                        ContractOutputs { structs: HashMap::new(), globals: HashMap::new() };
+
+                    interner.get_all_globals().iter().for_each(|global_info| {
+                        interner.global_attributes(&global_info.id).iter().for_each(|attr| {
+                            if let SecondaryAttribute::Abi(tag) = attr {
+                                if let Some(tagged) = outputs.globals.get_mut(tag) {
+                                    tagged.push(global_info.id);
+                                } else {
+                                    outputs.globals.insert(tag.to_string(), vec![global_info.id]);
+                                }
+                            }
+                        });
+                    });
+
+                    module.type_definitions().for_each(|id| {
+                        if let ModuleDefId::TypeId(struct_id) = id {
+                            interner.struct_attributes(&struct_id).iter().for_each(|attr| {
+                                if let SecondaryAttribute::Abi(tag) = attr {
+                                    if let Some(tagged) = outputs.structs.get_mut(tag) {
+                                        tagged.push(struct_id);
+                                    } else {
+                                        outputs.structs.insert(tag.to_string(), vec![struct_id]);
+                                    }
+                                }
+                            });
+                        }
+                    });
```

Comment:
Have you checked that this won't pull in any outputs from dependencies?

Reply:
We should probably restrict the usage of the `abi` attribute to contract crates seeing as they're only useful there. We can warn on bin types and error on libs.

Reply:
Unfortunately we actually need to export data from dependencies (for example, exporting the note_id from a `ValueNote`, which is imported from `aztec`.

Reply:
But you're not wanting to export `ValueNote` unconditionally for all contracts even if it doesn't use it right? We need to have some opt-in mechanism for exporting values from dependencies in that case.

Reply:
I think this is overoptimising for the aztec usecase here. We're going to get potentially huge artifact sizes just for specifying a dependency if we go down this route.

Reply:
I'd argue a badly optimized library is exactly the same case, devs must be wary of what they import. This is just a language feature that could be possibly abused, like any other. 

For aztec we're just exporting the bare minimum info we need from every note (type and id) and the storage layout. Both are "opt-in" since the user has to specify the `#aztec(note/storage)]` attribute on both (and are welcome to implement them themselves if they want to optimize)

Reply:
On a similar note, I believe the old `#[event]` attribute wasn't protected either, but I haven't checked.

Reply:
I found a workaround for the problem on the `aztec-macros` side and I can block this attribute from being used in non-contract programs. Stand by for updates...